File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/rust/f64_roundToInt.rs
#![allow(non_snake_case, unused_parens, unused_imports, unused_variables, unused_comparisons, clippy::all)]
use crate::helpers::*;


pub fn f64_roundToInt_rm1(v_arg: u64, v_arg1: bool) -> u64 {
    let v_i: u64 = v_arg.wrapping_shr((0x34u64) as u32);
    let v_i2: u32 = ((v_i) as u32);
    let v_i3: u32 = (v_i2 & 0x7ffu32);
    let v_i4: bool = (v_i3 < 0x3ffu32);
    let v_i6: u64 = (v_arg & 0x7fffffffffffffffu64);
    let v_i7: bool = (v_i6 == 0x0u64);
    let v__m1: u64 = (0u64.wrapping_sub((v_i7) as u64));
    let v_i8: u64 = (v_arg & 0x8000000000000000u64);
    let v__a2: u64 = (v_arg & v__m1);
    let v__n3: u64 = (v__m1 ^ 0xffffffffffffffffu64);
    let v__a4: u64 = (v_i8 & v__n3);
    let v_spec_select: u64 = (v__a2 | v__a4);
    let v_i10: bool = (v_i3 > 0x432u32);
    let v_i12: bool = (v_i3 != 0x7ffu32);
    let v_i17: u32 = 0x433u32.wrapping_sub(v_i3);
    let v_i18: u64 = ((v_i17) as u64);
    let v__sh9: u64 = 0xffffffffffffffffu64.wrapping_shl((v_i18) as u32);
    let v__neg: u64 = v__sh9;
    let v_i19: u64 = (v_arg & v__neg);
    let v_i13: u64 = (v_arg & 0xfffffffffffffu64);
    let v_i14: bool = (v_i13 == 0x0u64);
    let v_i15: bool = (v_i14 | v_i12);
    let v__m5: u64 = (0u64.wrapping_sub((v_i15) as u64));
    let v__a6: u64 = (v_arg & v__m5);
    let v__n7: u64 = (v__m5 ^ 0xffffffffffffffffu64);
    let v__a8: u64 = (0x7ff8000000000000u64 & v__n7);
    let v_spec_select4: u64 = (v__a6 | v__a8);
    let v__n10: bool = (v_i10 ^ true);
    let v__n11: bool = (v_i4 ^ true);
    let v__c12: bool = (v__n10 & v__n11);
    let v__m13: u64 = (0u64.wrapping_sub((v__c12) as u64));
    let v__a14: u64 = (v_i19 & v__m13);
    let v__m15: u64 = (0u64.wrapping_sub((v_i4) as u64));
    let v__a16: u64 = (v_spec_select & v__m15);
    let v__o17: u64 = (v__a14 | v__a16);
    let v__m19: u64 = (0u64.wrapping_sub((v_i10) as u64));
    let v__a20: u64 = (v_spec_select4 & v__m19);
    let v__o21: u64 = (v__o17 | v__a20);
    v__o21
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/js/f64_roundToInt.js
'use strict';
const { sfUdiv, sfSgn, sfCtlz, sfAbs, sfUsubsat, sfFshl } = require('./helpers.js');


function f64_roundToInt_rm1(v_arg, v_arg1) {
  const v_i = (v_arg >> (0x34n & 0x3fn));
  const v_i2 = (v_i & 0xffffffffn);
  const v_i3 = (v_i2 & 0x7ffn);
  const v_i4 = ((v_i3 < 0x3ffn) ? 1n : 0n);
  const v_i6 = (v_arg & 0x7fffffffffffffffn);
  const v_i7 = ((v_i6 === 0x0n) ? 1n : 0n);
  const v__m1 = (sfSgn(v_i7, 0x1n) & 0xffffffffffffffffn);
  const v_i8 = (v_arg & 0x8000000000000000n);
  const v__a2 = (v_arg & v__m1);
  const v__n3 = (v__m1 ^ 0xffffffffffffffffn);
  const v__a4 = (v_i8 & v__n3);
  const v_spec_select = (v__a2 | v__a4);
  const v_i10 = ((v_i3 > 0x432n) ? 1n : 0n);
  const v_i12 = ((v_i3 !== 0x7ffn) ? 1n : 0n);
  const v_i17 = ((0x433n - v_i3) & 0xffffffffn);
  const v_i18 = (v_i17);
  const v__sh9 = ((0xffffffffffffffffn << (v_i18 & 0x3fn)) & 0xffffffffffffffffn);
  const v__neg = v__sh9;
  const v_i19 = (v_arg & v__neg);
  const v_i13 = (v_arg & 0xfffffffffffffn);
  const v_i14 = ((v_i13 === 0x0n) ? 1n : 0n);
  const v_i15 = (v_i14 | v_i12);
  const v__m5 = (sfSgn(v_i15, 0x1n) & 0xffffffffffffffffn);
  const v__a6 = (v_arg & v__m5);
  const v__n7 = (v__m5 ^ 0xffffffffffffffffn);
  const v__a8 = (0x7ff8000000000000n & v__n7);
  const v_spec_select4 = (v__a6 | v__a8);
  const v__n10 = (v_i10 ^ 0x1n);
  const v__n11 = (v_i4 ^ 0x1n);
  const v__c12 = (v__n10 & v__n11);
  const v__m13 = (sfSgn(v__c12, 0x1n) & 0xffffffffffffffffn);
  const v__a14 = (v_i19 & v__m13);
  const v__m15 = (sfSgn(v_i4, 0x1n) & 0xffffffffffffffffn);
  const v__a16 = (v_spec_select & v__m15);
  const v__o17 = (v__a14 | v__a16);
  const v__m19 = (sfSgn(v_i10, 0x1n) & 0xffffffffffffffffn);
  const v__a20 = (v_spec_select4 & v__m19);
  const v__o21 = (v__o17 | v__a20);
  return v__o21;
}
module.exports = { f64_roundToInt_rm1 };
